optuna/samplers/__init__.py

```diff
@@ -1,64 +1,7 @@
-from collections import OrderedDict
-
-import optuna
+from optuna.samplers._search_space import intersection_search_space  # NOQA
+from optuna.samplers._search_space import IntersectionSearchSpace  # NOQA
 from optuna.samplers.base import BaseSampler  # NOQA
 from optuna.samplers.cmaes import CmaEsSampler  # NOQA
 from optuna.samplers.grid import GridSampler  # NOQA
 from optuna.samplers.random import RandomSampler  # NOQA
 from optuna.samplers.tpe import TPESampler  # NOQA
-
-if optuna.type_checking.TYPE_CHECKING:
-    from typing import Dict  # NOQA
-
-    from optuna.distributions import BaseDistribution  # NOQA
-    from optuna.study import BaseStudy  # NOQA
-
-
-def intersection_search_space(study, ordered_dict=False):
-    # type: (BaseStudy, bool) -> Dict[str, BaseDistribution]
-    """Return the intersection search space of the :class:`~optuna.study.BaseStudy`.
-
-    Intersection search space contains the intersection of parameter distributions that have been
-    suggested in the completed trials of the study so far.
-    If there are multiple parameters that have the same name but different distributions,
-    neither is included in the resulting search space
-    (i.e., the parameters with dynamic value ranges are excluded).
-
-    Args:
-        study:
-            A study with completed trials.
-        ordered_dict:
-            A boolean flag determining the return type.
-            If :obj:`False`, the returned object will be a :obj:`dict`.
-            If :obj:`True`, the returned object will be an :obj:`collections.OrderedDict` sorted by
-            keys, i.e. parameter names.
-
-    Returns:
-        A dictionary containing the parameter names and parameter's distributions.
-    """
-
-    search_space = None
-    for trial in study.trials:
```

Comment:
Yes, you're right. The removal of the deep copy has much impact on this performance improvement.

If we only applied the removal to the `master` branch, the benchmark result become as follows:
```
$ time -p python3  /tmp/benchmark.py  --params 10 --trials 1000
{'0': 1.9531119708170848, '1': 2.086781403839324, '2': 2.0622082517919336, '3': 1.9866026063257591, '4': 2.058500085804301, '5': 2.020942555633829, '6': 1.9894401312961876, '7': 2.043186636871422, '8': 1.954916778325024, '9': 2.093560500426215}
real 11.45
user 18.65
sys 12.60
```

Diff from `master`:
```diff
diff --git a/optuna/samplers/__init__.py b/optuna/samplers/__init__.py
index 8daa7143..e6f636c8 100644
--- a/optuna/samplers/__init__.py
+++ b/optuna/samplers/__init__.py
@@ -1,4 +1,5 @@
 from collections import OrderedDict
+import copy
 
 import optuna
 from optuna.samplers.base import BaseSampler  # NOQA
@@ -38,12 +39,12 @@ def intersection_search_space(study, ordered_dict=False):
     """
 
     search_space = None
-    for trial in study.trials:
+    for trial in study.get_trials(deepcopy=False):
         if trial.state != optuna.trial.TrialState.COMPLETE:
             continue
 
         if search_space is None:
-            search_space = trial.distributions
+            search_space = copy.copy(trial.distributions)
             continue
 
         delete_list = []
@@ -61,4 +62,4 @@ def intersection_search_space(study, ordered_dict=False):
     if ordered_dict:
         search_space = OrderedDict(sorted(search_space.items(), key=lambda x: x[0]))
 
-    return search_space
+    return copy.deepcopy(search_space)
```

Summary:
- `master`: 96.11s
- Remove deepcopy: 11.45s
- Add search space cache: 7.30s

I think that it might be reasonable to only apply the deep copy removal as it requires only a few changes of the code.

Reply:
I think the speed gain overwhelms the cost of code changes. The reorganization is also nice.